docs/extensions/guides/ipc.md

```diff
@@ -0,0 +1,105 @@
+
+IPC channels are blocked off per extension.
+Meaning that each extension can only communicate with itself.
+
+## Types of IPC
+
+There are two flavours of IPC that are provided:
+
+- Event based
+- Request based
+
+### Event Based IPC
+
+This is the same as an [Event Emitter](https://nodejs.org/api/events.html#events_class_eventemitter) but is not limited to just one Javascript process.
+This is a good option when you need to report that something has happened but you don't need a response.
+
+This is a fully two-way form of communication.
+Both `LensMainExtension` and `LensRendererExtension` can do this sort of IPC.
+
+### Request Based IPC
+
+This is more like a Remote Procedure Call (RPC).
+With this sort of IPC the caller waits for the result from the other side.
+This is accomplished by returning a `Promise<T>` which needs to be `await`-ed.
+
+This is a unidirectional form of communication.
+Only `LensRendererExtension` can initiate this kind of request, and only `LensMainExtension` can and respond this this kind of request.
+
+## Registering IPC Handlers and Listeners
+
+The general terminology is as follows:
+
+- A "handler" is the function that responds to a "Request Based IPC" event.
+- A "listener" is the function that is called when a "Event Based IPC" event is emitted.
+
+To register either a handler or a listener, you should do something like the following:
+
+File1:
+```typescript
+import { LensMainExtension, Interface, Types } from "@k8slens/extensions";
+
+export class ExampleExtensionMain extends LensMainExtension {
+  onActivate() {
+    this.listenIpc({
+      channel: "initialize",
+      listener: this.initializeListener,
+      verifier: this.initializeVerifier,
+    });
+  }
+
+  initializeListener = (event: Types.IpcMainEvent, id: string) => {
+    console.log(`starting to initialize: ${id}`);
+  };
+
+  initializeVerifier = (args: unknown[]): args is [id: string] => {
+    return args.length === 1 && typeof args[0] === "string";
+  }
+}
+```
+
+File2:
+```typescript
+import { LensMainExtension, Interface, Types } from "@k8slens/extensions";
+
+export class ExampleExtensionRenderer extends LensRendererExtension {
+  onActivate() {
+    setTimeout(() => this.sendIpc("initialize", "an-id"), 5000);
+  }
+}
+```
+
+As this example shows: the channel names *must* be the same.
+It should also be noted that "listeners" and "handlers" are specific to either `LensRendererExtension` and `LensMainExtension`.
+There is no behind the scenes transfer of these functions.
+
+If you want to register a "handler" you would call `this.handleIpc(...)` instead.
+The cleanup of these handlers is handled by Lens itself.
+
+### Note about verification
+
+We require that extension developers provide a verification function when registering a listener or handler.
+This is done as a preventative measure to help separate issues that can happen at runtime.
+While it is possible to use the unary truth function, this is highly discouraged.
```

Comment:
I'm a big proponent of runtime verifications. I wrote [rtv.js](https://rtvjs.stefcameron.com/) after all, and we use it in our MDS Dev builds for that very purpose. So I get the logic of enforcing the verifier, but as much as I like runtime verification and its benefits, I don't think it should be enforced. I see this as a nice to have for that extra step of quality. But because there's nothing you can do to prevent an extension developer from just providing you with `() => true`, I think it'll be seen as strange to be required.

But if you're going to require it, then I think it would be nice to also support throwing an error to fail the validation, and then Lens can log the error as a result.

There are a number of runtime validation libraries people will likely want to use, like [yup](https://github.com/jquense/yup), [joi](https://github.com/sideway/joi), or maybe even [rtv.js](https://rtvjs.stefcameron.com/) 🙂 

If I were using `rtv.js` here, I'd find it handy to write this:

```javascript
  initializeVerifier = (args) => {
    rtv.verify(args, [rtv.ARRAY, { $: [rtv.STRING], length: 1 }]);
  }
```

And then let that blow-up if `args` is `[]` or `['']` or `['asdf', 3]` or not an array.

Reply:
So for some context this will work too as it stands since we currently only progress if the return value is truthy. We even already have a `try/catch` block and do log the error.

The concept of returning `true` here comes from Typescript's https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Reply:
I also think this should not be enforced. I get this kind of verification if listener/handler would have external consumers but now the consumer is the extension itself.